### PR TITLE
[sonic-cfggen]: Update UT to run yang validation

### PR DESCRIPTION
**Pull request by ganglyu**
Signed-off-by: Gang Lv ganglv@microsoft.com

#### Why I did it
Config db schema generated by minigraph should run yang validation.

#### How I did it
Modify run_script to add yang validation.

#### How to verify it
Run sonic-config-engine unit test.

#### Which release branch to backport (provide reason below if selected)

- [ ] 201811
- [ ] 201911
- [ ] 202006
- [ ] 202012
- [ ] 202106

### CODE REVIEW

File: src/sonic-config-engine/tests/test_cfggen.py [hunk 1]
```diff
@@ -49,6 +59,32 @@ def tearDown(self):
 
     def run_script(self, argument, check_stderr=False, verbose=False):
         print('\n    Running sonic-cfggen ' + argument)
+        if PY3x and "-m" in argument:
```
Comment (qiluo-msft) [inline src/sonic-config-engine/tests/test_cfggen.py:62]: >argument [](http://example.com/codeflow?start=28&length=8)

Suggest use `ArgumentParser` and `shlex.split()` to parse the arguments in a string. ref: [https://stackoverflow.com/a/18596435/2514803](https://stackoverflow.com/a/18596435/2514803)

File: src/sonic-config-engine/tests/test_cfggen.py [hunk 2]
```diff
@@ -49,6 +59,32 @@ def tearDown(self):
...
+                cmd = self.script_file + ' -m ' + minigraph[0]
+                if hwsku:
+                    cmd += ' -S ' + hwsku[0]
+                if port_conf:
+                    cmd += ' -p ' + port_conf[0]
+                if namespace:
+                    cmd += ' -n ' + namespace[0]
+                cmd += ' --print-data'
+                output = subprocess.check_output(cmd, shell=True).decode()
+                # "NULL": "NULL" is placeholder for redis, remove to pass yang validation.
```
Comment (qiluo-msft) [inline src/sonic-config-engine/tests/test_cfggen.py:79]: >"NULL": "NULL" [](http://example.com/codeflow?start=18&length=14)

This seems a bug. I submit a PR to fix [https://github.com/Azure/sonic-buildimage/pull/9703](https://github.com/Azure/sonic-buildimage/pull/9703).

File: src/sonic-config-engine/tests/test_minigraph_case.py [hunk 3]
```diff
@@ -23,6 +33,32 @@ def setUp(self):
 
     def run_script(self, argument, check_stderr=False):
```
Comment (qiluo-msft) [inline src/sonic-config-engine/tests/test_minigraph_case.py:34]: >run_script [](http://example.com/codeflow?start=8&length=10)

Duplicated code. Could you reuse?

File: src/sonic-config-engine/tests/test_multinpu_cfggen.py [hunk 4]
```diff
@@ -33,6 +42,32 @@ def setUp(self):
 
     def run_script(self, argument, check_stderr=False):
```
Comment (qiluo-msft) [inline src/sonic-config-engine/tests/test_multinpu_cfggen.py:43]: >run_script [](http://example.com/codeflow?start=8&length=10)

Duplicated code. Could you reuse?

**Comment by qiluo-msft** [comment]
@judyjoseph Could you check this unit test improvement also working for multi-ASIC?

File: src/sonic-config-engine/tests/common_utils.py [hunk 5]
```diff
@@ -31,3 +37,46 @@ def liststr_to_dict(liststr):
...
+                cmd += ' -k ' + args.hwsku
+            if args.hwsku_config is not None:
+                cmd += ' -S ' + args.hwsku_config
+            if args.port_config is not None:
+                cmd += ' -p ' + args.port_config
+            if args.namespace is not None:
+                cmd += ' -n ' + args.namespace
+            cmd += ' --print-data'
+            output = subprocess.check_output(cmd, shell=True).decode()
+            self.yang_parser.loadData(configdbJson=json.loads(output))
```
Comment (ghooo) [inline src/sonic-config-engine/tests/common_utils.py:77]: This line can throw, place it in the try catch.

**Review by qiluo-msft — DISMISSED** [review]
LGTM. Please check with other reviewers.

File: src/sonic-config-engine/tests/common_utils.py [hunk 6]
```diff
@@ -31,3 +37,47 @@ def liststr_to_dict(liststr):
...
+            if args.namespace is not None:
+                cmd += ' -n ' + args.namespace
+            cmd += ' --print-data'
+            output = subprocess.check_output(cmd, shell=True).decode()
+            try:
+                self.yang_parser.loadData(configdbJson=json.loads(output))
+                self.yang_parser.validate_data_tree()
+            except sonic_yang.SonicYangException as e:
+                print("yang data generated from %s is not valid"%(args.minigraph))
+                raise
```
Comment (qiluo-msft) [inline src/sonic-config-engine/tests/common_utils.py:83]: >raise [](http://example.com/codeflow?start=16&length=5)

`raise` is not friendly in unit test, You may return a bool, and a string. And the caller could use `assert` to apply the pair.

Validation failure should be treated as a test case failure, not an exception.

**Review by qiluo-msft — DISMISSED** [review]

**Review by ghooo — DISMISSED** [review]

**Comment by zhangyanzhao** [comment]
Still has one blocking issue that MSFT team is working on.

**Comment by zhangyanzhao** [comment]
Still one more blocking issue  and MSFT team is working on that issue.

**Comment by zhangyanzhao** [comment]
This is addressed by https://github.com/Azure/sonic-buildimage/pull/10786 and close this PR.

**Comment by qiluo-msft** [comment]
This PR is still open.

**Review by qiluo-msft — APPROVED** [review]

**Comment by stephenxs** [comment]
@judyjoseph @qiluo-msft 
Looks like cherry-picking this PR to 202111 fails the build test. Can you please check whether we miss something other PRs?
```
2022-05-17T04:17:53.5225444Z ======================================================================
2022-05-17T04:17:53.5226298Z FAIL: test_minigraph_voq_metadata (tests.test_cfggen.TestCfgGen)
2022-05-17T04:17:53.5227843Z ----------------------------------------------------------------------
2022-05-17T04:17:53.5228597Z Traceback (most recent call last):
2022-05-17T04:17:53.5229846Z   File "/sonic/src/sonic-config-engine/tests/test_cfggen.py", line 827, in test_minigraph_voq_metadata
2022-05-17T04:17:53.5230781Z     output = json.loads(self.run_script(argument))
2022-05-17T04:17:53.5232119Z   File "/sonic/src/sonic-config-engine/tests/test_cfggen.py", line 54, in run_script
2022-05-17T04:17:53.5232954Z     self.assertTrue(self.yang.validate(argument))
2022-05-17T04:17:53.5233700Z AssertionError: False is not true
2022-05-17T04:17:53.5233988Z 
2022-05-17T04:17:53.5234579Z ======================================================================
2022-05-17T04:17:53.5235398Z FAIL: test_minigraph_voq_system_ports (tests.test_cfggen.TestCfgGen)
2022-05-17T04:17:53.5236814Z ----------------------------------------------------------------------
2022-05-17T04:17:53.5237538Z Traceback (most recent call last):
2022-05-17T04:17:53.5238882Z   File "/sonic/src/sonic-config-engine/tests/test_cfggen.py", line 836, in test_minigraph_voq_system_ports
2022-05-17T04:17:53.5240263Z     json.loads(self.run_script(argument)),
2022-05-17T04:17:53.5241540Z   File "/sonic/src/sonic-config-engine/tests/test_cfggen.py", line 54, in run_script
2022-05-17T04:17:53.5242388Z     self.assertTrue(self.yang.validate(argument))
2022-05-17T04:17:53.5243050Z AssertionError: False is not true
2022-05-17T04:17:53.5243783Z 
```
Check the build failure in #10843 for details.
Thanks

**Comment by qiluo-msft** [comment]
@ganglyu is working on reverting on 202111.

**Pull request merged**